Failing unit tests on ARM

**$** uname -a
Linux scw-c40323 4.1.6-249 #1 SMP Mon Aug 31 12:34:47 UTC 2015 **armv7l** GNU/Linux
**$** go version
go version go1.5.3 **linux/arm**
**$** go test github.com/btcsuite/btcd/...
\# github.com/btcsuite/btcd/btcjson_test
**go/src/github.com/btcsuite/btcd/btcjson/chainsvrcmds_test.go:66: constant 12312333333 overflows int**
ok      github.com/btcsuite/btcd        0.414s
ok      github.com/btcsuite/btcd/addrmgr        2.359s
ok      github.com/btcsuite/btcd/blockchain     7.117s
ok      github.com/btcsuite/btcd/btcec  35.714s
**FAIL    github.com/btcsuite/btcd/btcjson [build failed]**
ok      github.com/btcsuite/btcd/chaincfg       0.017s
?       github.com/btcsuite/btcd/cmd/addblock   [no test files]
?       github.com/btcsuite/btcd/cmd/btcctl     [no test files]
?       github.com/btcsuite/btcd/cmd/dropafter  [no test files]
?       github.com/btcsuite/btcd/cmd/findcheckpoint     [no test files]
?       github.com/btcsuite/btcd/cmd/gencerts   [no test files]
ok      github.com/btcsuite/btcd/database       2.787s
ok      github.com/btcsuite/btcd/database/ldb   1.038s
?       github.com/btcsuite/btcd/database/ldb/dbtest    [no test files]
ok      github.com/btcsuite/btcd/database/memdb 0.437s
?       github.com/btcsuite/btcd/limits [no test files]
?       github.com/btcsuite/btcd/mining [no test files]
ok      github.com/btcsuite/btcd/peer   0.463s
**--- FAIL: TestTxValidTests (0.52s)**
        reference_test.go:580: bad test (missing 0th input) 46:[[[0000000000000000000000000000000000000000000000000000000000000000 -1 1]] 01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff025151ffffffff010000000000000000015100000000 P2SH]
        reference_test.go:580: bad test (missing 0th input) 49:[[[0000000000000000000000000000000000000000000000000000000000000000 -1 1]] 01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff6451515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151515151ffffffff010000000000000000015100000000 P2SH]
--- FAIL: TestCalcSignatureHash (0.12s)
        reference_test.go:650: TestCalcSignatureHash failed test #2: Signature hash mismatch.
        reference_test.go:650: TestCalcSignatureHash failed test #5: Signature hash mismatch.
        reference_test.go:650: TestCalcSignatureHash failed test #9: Signature hash mismatch.
        [... snip 12, 14, 15, 17, 18, 21, 22, 24, 25, 26, 29, 30, 31, 32, 34, 37, 39, 41, 43, 45, 50, 52, 55, 57, 60, 63, 64, 65, 67, 68, 72, 73, 76, 83, 84, 85, 86, 88, 92, 93, 94, 97, 98, 100, 101, 102, 106, 107, 111, 112, 113, 114, 119, 121, 122, 123, 124, 126, 127, 130, 131, 132, 133, 136, 137, 138, 140, 141, 145, 148, 149, 153, 154, 156, 158, 159, 160, 162, 167, 168, 169, 171, 175, 176, 177, 179, 185, 186, 187, 189, 191, 192, 193, 200, 203, 204, 205, 206, 209, 210, 213, 214, 216, 222, 224, 229, 230, 235, 238, 239, 240, 241, 243, 244, 246, 247, 248, 249, 250, 253, 256, 259, 261, 262, 265, 266, 272, 273, 274, 275, 276, 277, 279, 281, 282, 285, 288, 289, 290, 291, 292, 295, 296, 297, 298, 299, 300, 301, 303, 305, 309, 311, 312, 314, 315, 318, 319, 323, 325, 326, 328, 330, 331, 333, 334, 336, 338, 339, 341, 343, 345, 346, 347, 350, 352, 354, 355, 356, 357, 361, 363, 365, 367, 370, 371, 372, 373, 375, 377, 378, 379, 380, 381, 383, 384, 387, 389, 392, 394, 396, 402, 403, 409, 413, 415, 417, 418, 419, 420, 421, 423, 426, 429, 432, 433, 434, 437, 438, 443, 445, 450, 451, 452, 453, 457, 458, 461, 465, 469, 470, 471, 472, 475, 476, 479, 480, 481, 483, 484, 485, 486, 488, 489, 490, 491, 492, 496...]
        reference_test.go:650: TestCalcSignatureHash failed test #498: Signature hash mismatch.
FAIL
FAIL    github.com/btcsuite/btcd/txscript       12.265s
ok      github.com/btcsuite/btcd/wire   2.019s

It fixes the first two issues, unfortunately, there are still signature hashes failing (the same ones as above).

ok      github.com/btcsuite/btcd        0.395s
ok      github.com/btcsuite/btcd/addrmgr        2.517s
ok      github.com/btcsuite/btcd/blockchain     7.522s
ok      github.com/btcsuite/btcd/btcec  36.963s
ok      github.com/btcsuite/btcd/btcjson        0.059s
ok      github.com/btcsuite/btcd/chaincfg       0.040s
?       github.com/btcsuite/btcd/cmd/addblock   [no test files]
?       github.com/btcsuite/btcd/cmd/btcctl     [no test files]
?       github.com/btcsuite/btcd/cmd/dropafter  [no test files]
?       github.com/btcsuite/btcd/cmd/findcheckpoint     [no test files]
?       github.com/btcsuite/btcd/cmd/gencerts   [no test files]
ok      github.com/btcsuite/btcd/database       2.762s
ok      github.com/btcsuite/btcd/database/ldb   1.076s
?       github.com/btcsuite/btcd/database/ldb/dbtest    [no test files]
ok      github.com/btcsuite/btcd/database/memdb 0.426s
?       github.com/btcsuite/btcd/limits [no test files]
?       github.com/btcsuite/btcd/mining [no test files]
ok      github.com/btcsuite/btcd/peer   0.484s
**--- FAIL: TestCalcSignatureHash (0.12s)**
        reference_test.go:660: TestCalcSignatureHash failed test #2: Signature hash mismatch.
        reference_test.go:660: TestCalcSignatureHash failed test #5: Signature hash mismatch.
        [... snip 9, 12, 14, 15, 17, 18, 21, 22, 24, 25, 26, 29, 30, 31, 32, 34, 37, 39, 41, 43, 45, 50, 52, 55, 57, 60, 63, 64, 65, 67, 68, 72, 73, 76, 83, 84, 85, 86, 88, 92, 93, 94, 97, 98, 100, 101, 102, 106, 107, 111, 112, 113, 114, 119, 121, 122, 123, 124, 126, 127, 130, 131, 132, 133, 136, 137, 138, 140, 141, 145, 148, 149, 153, 154, 156, 158, 159, 160, 162, 167, 168, 169, 171, 175, 176, 177, 179, 185, 186, 187, 189, 191, 192, 193, 200, 203, 204, 205, 206, 209, 210, 213, 214, 216, 222, 224, 229, 230, 235, 238, 239, 240, 241, 243, 244, 246, 247, 248, 249, 250, 253, 256, 259, 261, 262, 265, 266, 272, 273, 274, 275, 276, 277, 279, 281, 282, 285, 288, 289, 290, 291, 292, 295, 296, 297, 298, 299, 300, 301, 303, 305, 309, 311, 312, 314, 315, 318, 319, 323, 325, 326, 328, 330, 331, 333, 334, 336, 338, 339, 341, 343, 345, 346, 347, 350, 352, 354, 355, 356, 357, 361, 363, 365, 367, 370, 371, 372, 373, 375, 377, 378, 379, 380, 381, 383, 384, 387, 389, 392, 394, 396, 402, 403, 409, 413, 415, 417, 418, 419, 420, 421, 423, 426, 429, 432, 433, 434, 437, 438, 443, 445, 450, 451, 452, 453, 457, 458, 461, 465, 469, 470, 471, 472, 475, 476, 479, 480, 481, 483, 484, 485, 486, 488, 489, 490, 491 ...]
        reference_test.go:660: TestCalcSignatureHash failed test #492: Signature hash mismatch.
        reference_test.go:660: TestCalcSignatureHash failed test #496: Signature hash mismatch.
        reference_test.go:660: TestCalcSignatureHash failed test #498: Signature hash mismatch.
FAIL
FAIL    github.com/btcsuite/btcd/txscript       12.197s
ok      github.com/btcsuite/btcd/wire   2.005s



After replacing

    SigHashType(test[3].(float64)),

(at the bottom of reference_test.go) with

    SigHashType(int32(test[3].(float64))),

the tests pass.

Since all of these seem to only affect the tests, not the actual code, I'm going to build and deploy btcd, and we'll get to see if the network finds anything to complain about. Any particular options I should add to make sure it generates useful debug logs?

Thank you for all the great work on btcd, I'm really looking forward to see how it performs!

--- a/btcjson/chainsvrcmds_test.go
+++ b/btcjson/chainsvrcmds_test.go
@@ -63,7 +63,7 @@ func TestChainSvrCmds(t *testing.T) {
 			name: "createrawtransaction optional",
 			newCmd: func() (interface{}, error) {
 				return btcjson.NewCmd("createrawtransaction", `[{"txid":"123","vout":1}]`,
-					`{"456":0.0123}`, 12312333333)
+					`{"456":0.0123}`, int64(12312333333))
 			},
 			staticCmd: func() interface{} {
 				txInputs := []btcjson.TransactionInput{

--- a/txscript/reference_test.go
+++ b/txscript/reference_test.go
@@ -309,20 +309,31 @@ func TestScriptValidTests(t *testing.T) {
 	}
 }
 
+// testVecF64ToUint32 properly handles conversion of float64s read from the JSON
+// test data to unsigned 32-bit integers.  This is necessary because some of the
+// test data uses -1 as a shortcut to mean max uint32 and direct conversion of a
+// negative float to an unsigned int is implementation dependent and therefore
+// doesn't result in the expected value on all platforms.  This function woks
+// around that limitation by converting to a 32-bit signed integer first and
+// then to a 32-bit unsigned integer which results in the expected behavior on
+// all platforms.
+func testVecF64ToUint32(f float64) uint32 {
+	return uint32(int32(f))
+}
+
 // TestTxInvalidTests ensures all of the tests in tx_invalid.json fail as
 // expected.
 func TestTxInvalidTests(t *testing.T) {
 	file, err := ioutil.ReadFile("data/tx_invalid.json")
 	if err != nil {
-		t.Errorf("TestBitcoindInvalidTests: %v\n", err)
+		t.Errorf("TestTxInvalidTests: %v\n", err)
 		return
 	}
 
 	var tests [][]interface{}
 	err = json.Unmarshal(file, &tests)
 	if err != nil {
-		t.Errorf("TestBitcoindInvalidTests couldn't Unmarshal: %v\n",
-			err)
+		t.Errorf("TestTxInvalidTests couldn't Unmarshal: %v\n", err)
 		return
 	}
 
@@ -409,8 +420,7 @@ testloop:
 					"%d: %v", j, i, test)
 				continue testloop
 			}
-
-			idx := uint32(idxf) // (floor(idxf) == idxf?)
+			idx := testVecF64ToUint32(idxf)
 
 			oscript, ok := input[2].(string)
 			if !ok {
@@ -459,15 +469,14 @@ testloop:
 func TestTxValidTests(t *testing.T) {
 	file, err := ioutil.ReadFile("data/tx_valid.json")
 	if err != nil {
-		t.Errorf("TestBitcoindInvalidTests: %v\n", err)
+		t.Errorf("TestTxValidTests: %v\n", err)
 		return
 	}
 
 	var tests [][]interface{}
 	err = json.Unmarshal(file, &tests)
 	if err != nil {
-		t.Errorf("TestBitcoindInvalidTests couldn't Unmarshal: %v\n",
-			err)
+		t.Errorf("TestTxValidTests couldn't Unmarshal: %v\n", err)
 		return
 	}
 
@@ -553,8 +562,7 @@ testloop:
 					"%d: %v", j, i, test)
 				continue
 			}
-
-			idx := uint32(idxf) // (floor(idxf) == idxf?)
+			idx := testVecF64ToUint32(idxf)
 
 			oscript, ok := input[2].(string)
 			if !ok {
@@ -640,9 +648,10 @@ func TestCalcSignatureHash(t *testing.T) {
 				"Failed to parse sub-script: %v", i, err)
 			continue
 		}
-		hash := TstCalcSignatureHash(parsedScript,
-			SigHashType(test[3].(float64)),
-			tx, int(test[2].(float64)))
+
+		hashType := SigHashType(testVecF64ToUint32(test[3].(float64)))
+		hash := TstCalcSignatureHash(parsedScript, hashType, tx,
+			int(test[2].(float64)))
 
 		expectedHash, _ := wire.NewShaHashFromStr(test[4].(string))
 		if !bytes.Equal(hash, expectedHash.Bytes()) {

